File: LunarML/map.js
"use strict";
function _list(a) {
    let x = null;
    for (let i = a.length - 1; i >= 0; --i) {
        x = [a[i], x];
    }
    return x;
}
function _Match_tag() {}
_Match_tag.prototype.__isMLExn = true;
_Match_tag.prototype.name = "Match";
const _Match = new _Match_tag();
function _Overflow_tag() {}
_Overflow_tag.prototype.__isMLExn = true;
_Overflow_tag.prototype.name = "Overflow";
const _Overflow = new _Overflow_tag();
const MIN_INT54 = -0x20000000000000;
const MAX_INT54 = 0x1fffffffffffff;
function _Int54_add(x, y) {
    const z = x + y;
    if ((MIN_INT54 < z && z <= MAX_INT54) || (z === MIN_INT54 && (x & 1) === (y & 1))) {
        return z;
    } else {
        throw _Overflow;
    }
}
{
 let map;
 map = function(f) {
  return function(a) {
   if (a === null) {
    return null;
   } else if (! (a === null)) {
    const tmp1 = a[0];
    const tmp2 = a[1];
    const tmp3 = f(tmp1);
    const tmp4 = map(f);
    const tmp5 = tmp4(tmp2);
    return [tmp3, tmp5];
   } else {
    throw _Match;
   }
  };
 };
 const tmp = map(function(x) {
  return _Int54_add(x, 1);
 });
 tmp(_list([1, 2, 3]));
}

File: LunarML/map.lua
local error = error
local getmetatable = getmetatable
local setmetatable = setmetatable
local string = string
local string_format = string.format
local _exn_meta = {}
function _exn_meta:__tostring()
  local traceback = self.traceback
  if traceback then
    traceback = "\n" .. traceback
  else
    traceback = ""
  end
  return string_format("%s: %s%s", self.location or "<no location info>", self.tag[1], traceback)
end
local _Match_tag = { "Match" }
local _Match = setmetatable({ tag = _Match_tag }, _exn_meta)
local _Overflow_tag = { "Overflow" }
local _Overflow = setmetatable({ tag = _Overflow_tag }, _exn_meta)
local function _raise(x, location)
  local e
  if getmetatable(x) == _exn_meta and location ~= nil then
    local traceback = debug.traceback(nil, 2)
    e = setmetatable({ tag = x.tag, payload = x.payload, location = location, traceback = traceback }, _exn_meta)
  else
    e = x
  end
  error(e, 1)
end
local function _Int_add(x, y)
  -- assert(math_type(x) == "integer")
  -- assert(math_type(y) == "integer")
  local z = x + y
  if y > 0 and z < x then
    _raise(_Overflow, "Int.+")
  elseif y < 0 and z > x then
    _raise(_Overflow, "Int.+")
  else
    return z
  end
end
local function map(f)
  return function(a)
    if a == nil then
      return nil
    elseif a ~= nil then
      local tmp = a[1]
      local tmp1 = a[2]
      local tmp2 = f(tmp)
      local tmp3 = map(f)
      local tmp4 = tmp3(tmp1)
      return {tmp2, tmp4}
    else
      _raise(_Match, "map.sml:3:5")
    end
  end
end
local tmp = map(function(x)
  return _Int_add(x, 1)
end)
tmp({1, {2, {3, nil}}})
